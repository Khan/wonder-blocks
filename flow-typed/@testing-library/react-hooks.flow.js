// @flow
/**
 * This file is based on:
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/index.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/react.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/pure.ts
 */
 declare module "@testing-library/react-hooks" {
    declare export type Renderer<TProps> = {|
        render: (props?: TProps) => void,
        rerender: (props?: TProps) => void,
        unmount: () => void,
        act: Act,
    |};

    declare export type ServerRenderer<TProps> = {|
        ...Renderer<TProps>,
        hydrate: () => void,
    |};

    declare export type RendererProps<TProps, TResult> = {|
        callback: (props: TProps) => TResult,
        setError: (error: Error) => void,
        setValue: (value: TResult) => void,
    |};

    declare export type CreateRenderer<
        TProps,
        TResult,
        TRendererOptions: {...},
        TRenderer: Renderer<TProps>,
    > = (
        props: RendererProps<TProps, TResult>,
        options: TRendererOptions,
    ) => TRenderer;

    declare export type RenderResult<TValue> = {|
        +all: Array<TValue | Error>,
        +current: TValue,
        +error?: Error,
    |};

    declare export type ResultContainer<TValue> = {|
        result: RenderResult<TValue>,
    |};

    declare export type WaitOptions = {|
        interval?: number | false,
        timeout?: number | false,
    |};

    declare export type WaitForOptions = WaitOptions;
    declare export type WaitForValueToChangeOptions = WaitOptions;
    declare export type WaitForNextUpdateOptions = WaitOptions["timeout"];

    declare export type WaitFor = (
        callback: () => boolean | void,
        options?: WaitForOptions,
    ) => Promise<void>;
    declare export type WaitForValueToChange = (
        selector: () => mixed,
        options?: WaitForValueToChangeOptions,
    ) => Promise<void>;
    declare export type WaitForNextUpdate = (
        options?: WaitForNextUpdateOptions,
    ) => Promise<void>;

    declare export type AsyncUtils = {|
        waitFor: WaitFor,
        waitForValueToChange: WaitForValueToChange,
        waitForNextUpdate: WaitForNextUpdate,
    |};

    declare export type RenderHookResult<
        TProps,
        TValue,
        TRenderer: Renderer<TProps> = Renderer<TProps>,
    > = {|
        ...ResultContainer<TValue>,
        // Omit<Renderer<TProps>, 'render' | 'act'>
        ...$Diff<
            Renderer<TProps>,
            {|
                render: Renderer<TProps>["render"],
                act: Renderer<TProps>["act"],
            |},
        >,
        // Omit<TRenderer, keyof Renderer<TProps>>
        ...$Diff<TRenderer, Renderer<TProps>>,
        ...AsyncUtils,
    |};

    declare export type ServerRenderHookResult<
        TProps,
        TValue,
        TRenderer: ServerRenderer<TProps> = ServerRenderer<TProps>,
    > = RenderHookResult<TProps, TValue, TRenderer>;

    declare export type Act = {|
        (callback: () => Promise<void>): Promise<void>,
        (callback: () => void): void,
    |};

    declare export type CleanupCallback = () => Promise<void> | void;

    declare export type WrapperComponent<TProps> = React$ComponentType<TProps>;

    declare export type RendererOptions<TProps> = {|
        wrapper?: WrapperComponent<TProps>,
    |};

    declare export type RenderHookOptions<TProps> = {|
        initialProps?: TProps,
        wrapper?: WrapperComponent<any>,
    |};

    declare export type ReactHooksRenderer = {|
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => RenderHookResult<TProps, TResult>,
        act: Act,
        cleanup: () => Promise<void>,
        addCleanup: (callback: CleanupCallback) => () => void,
        removeCleanup: (callback: CleanupCallback) => void,
        suppressErrorOutput: () => () => void,
    |};

    declare export type ReactHooksServerRenderer = {|
        // Omit<ReactHooksRenderer, 'renderHook'>
        ...$Diff<
            ReactHooksRenderer,
            {|
                renderHook: ReactHooksRenderer["renderHook"],
            |},
        >,
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => ServerRenderHookResult<TProps, TResult>,
    |};

    declare export function renderHook<TProps, TResult>(
        callback: (props: TProps) => TResult,
        options?: RenderHookOptions<TProps>,
    ): RenderHookResult<TProps, TResult>;
    declare export var act: Act;
    declare export function cleanup(): Promise<void>;
    declare export function addCleanup(callback: CleanupCallback): () => void;
    declare export function removeCleanup(callback: CleanupCallback): void;
    declare export function suppressErrorOutput(): () => void;
}
