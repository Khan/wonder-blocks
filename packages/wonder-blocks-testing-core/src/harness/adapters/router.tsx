import * as React from "react";

import {StaticRouter, MemoryRouter, Route, Switch} from "react-router-dom";

import type {LocationDescriptor} from "history";
import type {TestHarnessAdapter} from "../types";

type MemoryRouterProps = JSX.LibraryManagedAttributes<
    typeof MemoryRouter,
    React.ComponentProps<typeof MemoryRouter>
>;

/**
 * Configuration for the withLocation test harness adapter.
 */
type Config = // The initial location to use.

        | Readonly<
              | {
                    /**
                     * See MemoryRouter prop for initialEntries.
                     */
                    initialEntries: MemoryRouterProps["initialEntries"];
                    /**
                     * See MemoryRouter prop for initialIndex.
                     */
                    initialIndex?: MemoryRouterProps["initialIndex"];
                    /**
                     * See MemoryRouter prop for getUserConfirmation.
                     */
                    getUserConfirmation?: MemoryRouterProps["getUserConfirmation"];
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
              | {
                    /**
                     * The location to use.
                     */
                    location: LocationDescriptor;
                    /**
                     * Force the use of a StaticRouter, instead of MemoryRouter.
                     */
                    forceStatic: true;
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
              | {
                    /**
                     * The initial location to use.
                     */
                    location: LocationDescriptor;
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
          >
        | string;

/**
 * The default configuration for this adapter.
 */
export const defaultConfig = {location: "/"} as const;

const maybeWithRoute = (
    children: React.ReactNode,
    path?: string | null,
): React.ReactElement => {
    if (path == null) {
        return <>{children}</>;
    }

    return (
        <Switch>
            <Route exact={true} path={path}>
                {children}
            </Route>
            <Route
                path="*"
                render={() => {
                    throw new Error(
                        "The configured path must match the configured location or your harnessed component will not render.",
                    );
                }}
            />
        </Switch>
    );
};

/**
 * Adapter that sets up a router and AppShell location-specific contexts.
 *
 * This allows you to ensure that components are being tested in the
 * AppShell world.
 *
 * NOTE(somewhatabstract): The AppShell component itself already does
 * the work of setting up routing and the AppShellContext and so using this
 * adapter with the App component will have zero-effect since AppShell will
 * override it.
 */
export const adapter: TestHarnessAdapter<Config> = (
    children: React.ReactNode,
    config: Config,
): React.ReactElement<any> => {
    if (typeof config === "string") {
        config = {
            location: config,
        };
    }

    // Wrap children with the various contexts and routes, as per the config.
    const wrappedWithRoute = maybeWithRoute(children, config.path);
    if ("forceStatic" in config && config.forceStatic) {
        /**
         * There may be times (SSR testing comes to mind) where we will be
         * really strict about not permitting client-side navigation events.
         */
        return (
            <StaticRouter location={config.location} context={{}}>
                {wrappedWithRoute}
            </StaticRouter>
        );
    }

    /**
     * OK, we must be OK with a memory router.
     *
     * There are two flavors of config for this. The easy one with just a
     * location, and the complex one for those gnarlier setups.
     *
     * First, the easy one.
     */
    if ("location" in config && config.location !== undefined) {
        return (
            <MemoryRouter initialEntries={[config.location]}>
                {wrappedWithRoute}
            </MemoryRouter>
        );
    }

    /**
     * If it's not the easy one, it should be the complex one.
     * Let's make sure we have good data (also keeps TypeScript happy).
     */
    if (!("initialEntries" in config) || config.initialEntries === undefined) {
        throw new Error(
            "A location or initial history entries must be provided.",
        );
    }

    /**
     * What should happen if no entries were in the array?
     * It likely uses the root one anyway, but a consistent API is what
     * we want, so let's ensure we always have our default location at least.
     */
    const entries =
        config.initialEntries.length === 0
            ? [defaultConfig.location]
            : config.initialEntries;

    // Memory router doesn't allow us to pass maybe types in its TypeScript types.
    // So let's build props then spread them.
    const routerProps: MemoryRouterProps = {
        initialEntries: entries,
    };
    if (config.initialIndex != null) {
        routerProps.initialIndex = config.initialIndex;
    }
    if (config.getUserConfirmation != null) {
        routerProps.getUserConfirmation = config.getUserConfirmation;
    }

    return <MemoryRouter {...routerProps}>{wrappedWithRoute}</MemoryRouter>;
};
