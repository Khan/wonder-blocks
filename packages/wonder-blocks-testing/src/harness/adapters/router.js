// @flow
import * as React from "react";

import {StaticRouter, MemoryRouter, Route, Switch} from "react-router-dom";

import type {LocationShape, Location} from "react-router-dom";
import type {TestHarnessAdapter} from "../types.js";

type MemoryRouterProps = React.ElementConfig<typeof MemoryRouter>;

/**
 * Configuration for the withLocation test harness adapter.
 */
type Config =
    | $ReadOnly<
          | {|
                /**
                 * See MemoryRouter prop for initialEntries.
                 */
                initialEntries: MemoryRouterProps["initialEntries"],

                /**
                 * See MemoryRouter prop for initialIndex.
                 */
                initialIndex?: MemoryRouterProps["initialIndex"],

                /**
                 * See MemoryRouter prop for getUserConfirmation.
                 */
                getUserConfirmation?: MemoryRouterProps["getUserConfirmation"],

                /**
                 * A path match to use.
                 *
                 * When this is specified, the harnessed component will be
                 * rendered inside a `Route` handler with this path.
                 *
                 * If the path matches the location, then the route will
                 * render the component.
                 *
                 * If the path does not match the location, then the route
                 * will not render the component.
                 */
                path?: string,
            |}
          | {|
                /**
                 * The location to use.
                 */
                location: string | Location,

                /**
                 * Force the use of a StaticRouter, instead of MemoryRouter.
                 */
                forceStatic: true,

                /**
                 * A path match to use.
                 *
                 * When this is specified, the harnessed component will be
                 * rendered inside a `Route` handler with this path.
                 *
                 * If the path matches the location, then the route will
                 * render the component.
                 *
                 * If the path does not match the location, then the route
                 * will not render the component.
                 */
                path?: string,
            |}
          | {|
                /**
                 * The initial location to use.
                 */
                location: string | LocationShape,

                /**
                 * A path match to use.
                 *
                 * When this is specified, the harnessed component will be
                 * rendered inside a `Route` handler with this path.
                 *
                 * If the path matches the location, then the route will
                 * render the component.
                 *
                 * If the path does not match the location, then the route
                 * will not render the component.
                 */
                path?: string,
            |},
      >
    // The initial location to use.
    | string;

/**
 * The default configuration for this adapter.
 */
export const defaultConfig = {location: "/"};

const maybeWithRoute = (children: React.Node, path: ?string): React.Node => {
    if (path == null) {
        return children;
    }

    return (
        <Switch>
            <Route exact={true} path={path}>
                {children}
            </Route>
            <Route
                path="*"
                render={() => {
                    throw new Error(
                        "The configured path must match the configured location or your harnessed component will not render.",
                    );
                }}
            />
        </Switch>
    );
};

/**
 * Adapter that sets up a router and AppShell location-specific contexts.
 *
 * This allows you to ensure that components are being tested in the
 * AppShell world.
 *
 * NOTE(somewhatabstract): The AppShell component itself already does
 * the work of setting up routing and the AppShellContext and so using this
 * adapter with the App component will have zero-effect since AppShell will
 * override it.
 */
export const adapter: TestHarnessAdapter<Config> = (
    children: React.Node,
    config: Config,
): React.Element<any> => {
    if (typeof config === "string") {
        config = {
            location: config,
        };
    }

    // Wrap children with the various contexts and routes, as per the config.
    const wrappedWithRoute = maybeWithRoute(children, config.path);
    if (config.forceStatic) {
        /**
         * There may be times (SSR testing comes to mind) where we will be
         * really strict about not permitting client-side navigation events.
         */
        return (
            <StaticRouter location={config.location} context={{}}>
                {wrappedWithRoute}
            </StaticRouter>
        );
    }

    /**
     * OK, we must be OK with a memory router.
     *
     * There are two flavors of config for this. The easy one with just a
     * location, and the complex one for those gnarlier setups.
     *
     * First, the easy one.
     */
    if (typeof config.location !== "undefined") {
        return (
            <MemoryRouter initialEntries={[config.location]}>
                {wrappedWithRoute}
            </MemoryRouter>
        );
    }

    /**
     * If it's not the easy one, it should be the complex one.
     * Let's make sure we have good data (also keeps flow happy).
     */
    if (typeof config.initialEntries === "undefined") {
        throw new Error(
            "A location or initial history entries must be provided.",
        );
    }

    /**
     * What should happen if no entries were in the array?
     * It likely uses the root one anyway, but a consistent API is what
     * we want, so let's ensure we always have our default location at least.
     */
    const entries =
        config.initialEntries.length === 0
            ? [defaultConfig.location]
            : config.initialEntries;

    // Memory router doesn't allow us to pass maybe types in its flow types.
    // So let's build props then spread them.
    const routerProps: MemoryRouterProps = {
        initialEntries: entries,
    };
    if (config.initialIndex != null) {
        routerProps.initialIndex = config.initialIndex;
    }
    if (config.getUserConfirmation != null) {
        routerProps.getUserConfirmation = config.getUserConfirmation;
    }

    return <MemoryRouter {...routerProps}>{wrappedWithRoute}</MemoryRouter>;
};
