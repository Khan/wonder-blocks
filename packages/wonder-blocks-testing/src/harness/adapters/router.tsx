import * as React from "react";

import {StaticRouter, MemoryRouter, Route, Switch} from "react-router-dom";

// @ts-expect-error [FEI-5019] - TS2305 - Module '"react-router-dom"' has no exported member 'LocationShape'. | TS2305 - Module '"react-router-dom"' has no exported member 'Location'.
import type {LocationShape, Location} from "react-router-dom";
import type {TestHarnessAdapter} from "../types";

type MemoryRouterProps = JSX.LibraryManagedAttributes<
    typeof MemoryRouter,
    React.ComponentProps<typeof MemoryRouter>
>;

/**
 * Configuration for the withLocation test harness adapter.
 */
type Config = // The initial location to use.

        | Readonly<
              | {
                    /**
                     * See MemoryRouter prop for initialEntries.
                     */
                    initialEntries: MemoryRouterProps["initialEntries"];
                    /**
                     * See MemoryRouter prop for initialIndex.
                     */
                    initialIndex?: MemoryRouterProps["initialIndex"];
                    /**
                     * See MemoryRouter prop for getUserConfirmation.
                     */
                    getUserConfirmation?: MemoryRouterProps["getUserConfirmation"];
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
              | {
                    /**
                     * The location to use.
                     */
                    location: string | Location;
                    /**
                     * Force the use of a StaticRouter, instead of MemoryRouter.
                     */
                    forceStatic: true;
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
              | {
                    /**
                     * The initial location to use.
                     */
                    location: string | LocationShape;
                    /**
                     * A path match to use.
                     *
                     * When this is specified, the harnessed component will be
                     * rendered inside a `Route` handler with this path.
                     *
                     * If the path matches the location, then the route will
                     * render the component.
                     *
                     * If the path does not match the location, then the route
                     * will not render the component.
                     */
                    path?: string;
                }
          >
        | string;

/**
 * The default configuration for this adapter.
 */
export const defaultConfig = {location: "/"} as const;

const maybeWithRoute = (
    children: React.ReactNode,
    path?: string | null,
): React.ReactElement => {
    if (path == null) {
        // @ts-expect-error [FEI-5019] - TS2322 - Type 'ReactNode' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.
        return children;
    }

    return (
        <Switch>
            <Route exact={true} path={path}>
                {children}
            </Route>
            <Route
                path="*"
                render={() => {
                    throw new Error(
                        "The configured path must match the configured location or your harnessed component will not render.",
                    );
                }}
            />
        </Switch>
    );
};

/**
 * Adapter that sets up a router and AppShell location-specific contexts.
 *
 * This allows you to ensure that components are being tested in the
 * AppShell world.
 *
 * NOTE(somewhatabstract): The AppShell component itself already does
 * the work of setting up routing and the AppShellContext and so using this
 * adapter with the App component will have zero-effect since AppShell will
 * override it.
 */
export const adapter: TestHarnessAdapter<Config> = (
    children: React.ReactNode,
    config: Config,
): React.ReactElement<any> => {
    if (typeof config === "string") {
        config = {
            location: config,
        };
    }

    // Wrap children with the various contexts and routes, as per the config.
    const wrappedWithRoute = maybeWithRoute(children, config.path);
    // @ts-expect-error [FEI-5019] - TS2339 - Property 'forceStatic' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
    if (config.forceStatic) {
        /**
         * There may be times (SSR testing comes to mind) where we will be
         * really strict about not permitting client-side navigation events.
         */
        return (
            // @ts-expect-error [FEI-5019] - TS2339 - Property 'location' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
            <StaticRouter location={config.location} context={{}}>
                {wrappedWithRoute}
            </StaticRouter>
        );
    }

    /**
     * OK, we must be OK with a memory router.
     *
     * There are two flavors of config for this. The easy one with just a
     * location, and the complex one for those gnarlier setups.
     *
     * First, the easy one.
     */
    // @ts-expect-error [FEI-5019] - TS2339 - Property 'location' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
    if (typeof config.location !== "undefined") {
        return (
            // @ts-expect-error [FEI-5019] - TS2339 - Property 'location' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
            <MemoryRouter initialEntries={[config.location]}>
                {wrappedWithRoute}
            </MemoryRouter>
        );
    }

    /**
     * If it's not the easy one, it should be the complex one.
     * Let's make sure we have good data (also keeps TypeScript happy).
     */
    // @ts-expect-error [FEI-5019] - TS2339 - Property 'initialEntries' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
    if (typeof config.initialEntries === "undefined") {
        throw new Error(
            "A location or initial history entries must be provided.",
        );
    }

    /**
     * What should happen if no entries were in the array?
     * It likely uses the root one anyway, but a consistent API is what
     * we want, so let's ensure we always have our default location at least.
     */
    const entries =
        // @ts-expect-error [FEI-5019] - TS2339 - Property 'initialEntries' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
        config.initialEntries.length === 0
            ? [defaultConfig.location]
            : // @ts-expect-error [FEI-5019] - TS2339 - Property 'initialEntries' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
              config.initialEntries;

    // Memory router doesn't allow us to pass maybe types in its TypeScript types.
    // So let's build props then spread them.
    const routerProps: MemoryRouterProps = {
        initialEntries: entries,
    };
    // @ts-expect-error [FEI-5019] - TS2339 - Property 'initialIndex' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
    if (config.initialIndex != null) {
        // @ts-expect-error [FEI-5019] - TS2339 - Property 'initialIndex' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
        routerProps.initialIndex = config.initialIndex;
    }
    // @ts-expect-error [FEI-5019] - TS2339 - Property 'getUserConfirmation' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
    if (config.getUserConfirmation != null) {
        // @ts-expect-error [FEI-5019] - TS2339 - Property 'getUserConfirmation' does not exist on type 'Readonly<{ initialEntries: LocationDescriptor<unknown>[] | undefined; initialIndex?: number | undefined; getUserConfirmation?: ((message: string, callback: (ok: boolean) => void) => void) | undefined; path?: string | undefined; } | { ...; } | { ...; }>'.
        routerProps.getUserConfirmation = config.getUserConfirmation;
    }

    return <MemoryRouter {...routerProps}>{wrappedWithRoute}</MemoryRouter>;
};
