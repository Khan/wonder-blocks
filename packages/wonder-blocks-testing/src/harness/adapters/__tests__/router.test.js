// @flow
import * as React from "react";
import {withRouter} from "react-router-dom";
import {render} from "@testing-library/react";
import * as Router from "../router.js";

describe("Router.adapter", () => {
    describe.each`
        type          | config
        ${"string"}   | ${"/math"}
        ${"location"} | ${{location: "/math"}}
        ${"full"}     | ${{initialEntries: ["/math"]}}
    `("with $type config", ({config}) => {
        it("should allow navigation", () => {
            // Arrange
            const historyListen = jest.fn();
            const HistoryListener = withRouter(({history}): React.Node => {
                React.useEffect(() => history.listen(historyListen), [history]);
                if (history.location.pathname === "/math") {
                    history.push("/math/calculator");
                }
                return null;
            });

            // Act
            render(Router.adapter(<HistoryListener />, config));

            // Assert
            expect(historyListen).not.toHaveBeenCalled();
        });

        it("should have default route match of root /", () => {
            // Arrange
            const matchCatcherFn = jest.fn();
            const MatchCatcher = withRouter(({match, history}): React.Node => {
                React.useEffect(() => {
                    if (history.location.pathname === "/math") {
                        history.push("/math/calculator");
                    }
                    matchCatcherFn(match);
                }, [match, history]);
                return null;
            });

            // Act
            render(Router.adapter(<MatchCatcher />, config));

            // Assert
            expect(matchCatcherFn).toHaveBeenLastCalledWith(
                expect.objectContaining({
                    path: "/",
                    url: "/",
                }),
            );
        });
    });

    describe.each`
        type          | config
        ${"location"} | ${{location: "/math/calculator", path: "/math/*"}}
        ${"full"}     | ${{initialEntries: ["/math/calculator"], path: "/math/*"}}
    `("with $type config including path", ({config}) => {
        it("should include routing for the given path", () => {
            // Arrange
            const matchCatcherFn = jest.fn();
            const MatchCatcher = withRouter(({match}): React.Node => {
                React.useEffect(() => {
                    matchCatcherFn(match);
                }, [match]);
                return null;
            });

            // Act
            render(Router.adapter(<MatchCatcher />, config));

            // Assert
            expect(matchCatcherFn).toHaveBeenLastCalledWith(
                expect.objectContaining({
                    isExact: true,
                    path: "/math/*",
                    url: "/math/calculator",
                }),
            );
        });

        it("should throw if the path does not match the location", () => {
            // Arrange
            // This is going to cause an error to be logged, so let's silence
            // that.
            jest.spyOn(console, "error").mockImplementation(() => {});
            const badConfig = {
                ...config,
                path: "/something/else/entirely",
            };

            // Act
            const underTest = () =>
                render(Router.adapter("CHILDREN", badConfig));

            // Assert
            expect(underTest).toThrowErrorMatchingSnapshot();
        });
    });

    describe("with forceStatic", () => {
        it("should not navigate", () => {
            // Arrange
            const historyListen = jest.fn();
            const HistoryListener = withRouter(({history}): React.Node => {
                React.useEffect(() => history.listen(historyListen), [history]);
                if (history.location.pathname === "/math") {
                    history.push("/math/calculator");
                }
                return null;
            });

            // Act
            render(
                Router.adapter(<HistoryListener />, {
                    location: "/math",
                    forceStatic: true,
                }),
            );

            // Assert
            expect(historyListen).not.toHaveBeenCalled();
        });
    });
});
