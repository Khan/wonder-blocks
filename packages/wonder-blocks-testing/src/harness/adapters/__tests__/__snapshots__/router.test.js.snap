// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router.adapter with full config including path should throw if the path does not match the location 1`] = `"The configured path must match the configured location or your harnessed component will not render."`;

exports[`Router.adapter with location config including path should throw if the path does not match the location 1`] = `"The configured path must match the configured location or your harnessed component will not render."`;
